### PR TITLE
Feat: DynamicAnnotation project type Backend (models, database structure, new API endpoints)

### PR DESCRIPTION
**Pull request by haanh764**
Hello, this PR aimed to create new project type `DynamicAnnotation` and some features for the backend side.

What's new:
- `backend/projects/models.py`: new project type models `DynamicAnnotation` project, new models for this project requirements - `DynamicDimension` (contains all dimensions), `ProjectDimension` (contains project and the dimensions assigned to each project), `DimensionMetaData` (contains metadata for each dimension)
- `backend/projects/migrations/0009_dynamicannotationproject_init_database.py`: new migration file to initialize default dimensions data for `DynamicDimension` table
- `backend/projects/migrations/0010_dynamicannotationproject_and_more.py`: new migration file to populate metadata for default dimensions for table `DimensionMetaData`
- `backend/projects/migrations/0011_dynamicannotationproject_and_more.py`: new migration file for new models
- `backend/projects/serializers.py`: added serializer classes for creating new project type and new dimension, serializers foe new models
- `backend/projects/views/dimension.py`: added APIs for create new dimensions, and list all dimensions, dimension's metadata
- `backend/projects/urls.py`: added new endpoints for new APIs
- `backend/projects/tests/test_project.py`: added test for new project type
- `backend/projects/tests/utils.py`: added new project type for test utils
- `backend/data_import/pipeline/catalog.py`: adjusted data import catalog for new project type
- `backend/data_import/datasets.py`: added new data import dataset for new project type
- `backend/data_import/pipeline/examples/dynamic_annotation/example.json`: added new data import example for new project type
- `backend/data_export/pipeline/factories.py`: adjusted data export factories for new project type
- `backend/data_export/pipeline/catalog.py`: adjusted data export catalog for new project type
- `backend/data_export/pipeline/examples/dynamic_annotation/`: added examples for data export in new project type

Endpoint description:
1. Create new project: the endpoint is the same as the other project type. The json format  for POST is as below:
```
{   
    "name": "dynamic project test",
    "project_type": "DynamicAnnotation",
    "description": "test dynamic project annotation proj",
    "dimension": [{"dimension": [2]}, {"dimension": [13]}, {"dimension": [15]}, {"dimension": [63]}],
    "resourcetype": "DynamicAnnotationProject"
}
```
`dimension` is the list of given dimension that will be assigned for this project. When a new DynamicAnnotation project is created, the corresponding information of `(project, dimensions)` will be inserted in table `ProjectDimension` as below:
Screenshot from 2023-03-13 17-21-09

2. Create new dimension
Endpoint: `/v1/projects/<int:project_id>/dimensions` (POST request)
Format of json:
```
{   
    "name": "New slider dimension test",
    "type": "slider",
    "dimension_meta_data": [{
        "codename": "DIM_NEW_EMO_0123",
        "config": {
            "slider_min": 0,
            "slider_max": 100,
            "slider_step": 10,
            "with_checkbox": 0,
            "minVal_description": "Wcale",
            "maxVal_description": "W pełni"
        },
        "required": true,
        "readonly": false}]
}
```
The response is as below (with the dimension id as required and other information as well)
Screenshot from 2023-03-13 17-25-53
When a new dimension is created, a new row is added in `DynamicDimension` table; a new row for metadata is inserted in `DimensionMetaData` table; the new created dimension will be assigned to the current project as well; and if the new dimension has `type: slider` a new scale label is added in `LabelType_ScaleType` table as below:

- New dimension created:
Screenshot from 2023-03-13 17-36-03

- New dimension's metadata inserted:
Screenshot from 2023-03-13 17-36-13

- Assigned new dimension to current project:
Screenshot from 2023-03-13 17-36-40

- Create new ScaleLabel for the new created dimension:
Screenshot from 2023-03-13 17-29-37

P/s: Currently, the slider dimension is corresponding to Scale label type, text field and multiple choice/single choice dimension are consider as categories which doesn't need to create label. That's why only when dimension's type is slider, a new label is created for current project. 

3. Additionally, just in case FE need in the future, there're 2 additional endpoints:
- `/v1/projects/dimensions/<int:dimension_id>/metadata` - to get the metadata of given dimension
- (GET method) `/v1/projects/dimensions` - list of all dimensions in `DynamicDimension` table

4. Assign dimensions from list of default dimensions to created project (in case a project was creating without given any dimension or admin want to add more dimensions after project created, there is ability to add the dimensions later using this endpoint):
Endpoint: `/v1/projects/<int:project_id>/assign_dimensions`
Format: 
```
{
    "dimension": [24, 25, 27]
}
```
Result will be: 
```
{
    "project": 1,
    "dimension": [
        24,
        25,
        27
    ]
}
```
Update: this endpoint will check if the selected dimension is default or no, and if it's not default dimension and is slider type -> new label associated with the dimension is added for the current project

### CODE REVIEW

**Comment by teddy-f-47** [comment]
> When a new dimension is created, a new row is added in DynamicDimension table; a new row for metadata is inserted in DimensionMetaData table; the new created dimension will be assigned to the current project as well; and if the new dimension has type: slider a new scale label is added in LabelType_ScaleType table as below:

Hello, thank you for the great work! I will have a deeper look on it, but I have a question about the creation of the scale label. Here, you described that when a new "dimension" with the type slider is created, it also automatically creates a "scale label" for the slider, and the "scale label" is associated with the current project. So, when the current project tries to list the scale labels, the scale label for the new dim will also be included in the list.

However, how about the case when we create a new project, and we select existing dimensions for it? Will this new project get associated with the "scale labels" of those existing dims?

File: backend/projects/views/dimension.py
Comment (teddy-f-47) [inline backend/projects/views/dimension.py:55]: In addition to my question previously, I have a suggestion here to add a check whether the given `project_id` and `dimension` pair already exists in the table. Because when I tested assign dimensions [1,2,3] to a project twice, the first time it, works but the second time it throws 500 error saying that `Key (project_id, dimension_id)=(136, 1) already exists` 😅 

Other than that, everything else seems to be working well :)

**Comment by haanh764** [comment]
> > When a new dimension is created, a new row is added in DynamicDimension table; a new row for metadata is inserted in DimensionMetaData table; the new created dimension will be assigned to the current project as well; and if the new dimension has type: slider a new scale label is added in LabelType_ScaleType table as below:
> 
> Hello, thank you for the great work! I will have a deeper look on it, but I have a question about the creation of the scale label. Here, you described that when a new "dimension" with the type slider is created, it also automatically creates a "scale label" for the slider, and the "scale label" is associated with the current project. So, when the current project tries to list the scale labels, the scale label for the new dim will also be included in the list.
> 
> However, how about the case when we create a new project, and we select existing dimensions for it? Will this new project get associated with the "scale labels" of those existing dims?

Hello, about this I think the question should be how FE-side maps the dimensions and the labels, right? For the default dimensions, there're default label configs which are kept in folder `/static` in FE side, and they are automatically added for a new project created. So in my opinion, we can just keep that mechanism and then all the project at starting will have all default label configs (just we wont use all of them depending on which dimensions are shown).

**Comment by teddy-f-47** [comment]
> > > When a new dimension is created, a new row is added in DynamicDimension table; a new row for metadata is inserted in DimensionMetaData table; the new created dimension will be assigned to the current project as well; and if the new dimension has type: slider a new scale label is added in LabelType_ScaleType table as below:
> > 
> > 
> > Hello, thank you for the great work! I will have a deeper look on it, but I have a question about the creation of the scale label. Here, you described that when a new "dimension" with the type slider is created, it also automatically creates a "scale label" for the slider, and the "scale label" is associated with the current project. So, when the current project tries to list the scale labels, the scale label for the new dim will also be included in the list.
> > However, how about the case when we create a new project, and we select existing dimensions for it? Will this new project get associated with the "scale labels" of those existing dims?
> 
> Hello, about this I think the question should be how FE-side maps the dimensions and the labels, right? For the default dimensions, there're default label configs which are kept in folder `/static` in FE side, and they are automatically added for a new project created. So in my opinion, we can just keep that mechanism and then all the project at starting will have all default label configs (just we wont use all of them depending on which dimensions are shown).

I mean for dimensions that are not default. For example, I have created a new dimension called "ABC" with type slider for project1. ABC does not exist in the list of default dimensions. And now I want to assign dimension ABC to a newly created project2. While we can assign ABC to project2, currently there is no way to add the scale label of ABC to project2. Though it is possible to simply create a new dimension in project2 with the same name, but then we will have two ABC dimensions in db (I thought the idea was that to re-use dimensions that are already created instead of making new dims with same names in every project)

File: backend/projects/views/dimension.py
Comment (haanh764) [reply backend/projects/views/dimension.py:55]: yeah, it is because of the unique constraint in Meta class of the model. It is not allowed to assign duplicated dimensions to the same project. So you will get the message of violate constraint. So I dont know what should I add here :sweat_smile: like a message or something?

**Comment by haanh764** [comment]
> > > > When a new dimension is created, a new row is added in DynamicDimension table; a new row for metadata is inserted in DimensionMetaData table; the new created dimension will be assigned to the current project as well; and if the new dimension has type: slider a new scale label is added in LabelType_ScaleType table as below:
> > > 
> > > 
> > > Hello, thank you for the great work! I will have a deeper look on it, but I have a question about the creation of the scale label. Here, you described that when a new "dimension" with the type slider is created, it also automatically creates a "scale label" for the slider, and the "scale label" is associated with the current project. So, when the current project tries to list the scale labels, the scale label for the new dim will also be included in the list.
> > > However, how about the case when we create a new project, and we select existing dimensions for it? Will this new project get associated with the "scale labels" of those existing dims?
> > 
> > 
> > Hello, about this I think the question should be how FE-side maps the dimensions and the labels, right? For the default dimensions, there're default label configs which are kept in folder `/static` in FE side, and they are automatically added for a new project created. So in my opinion, we can just keep that mechanism and then all the project at starting will have all default label configs (just we wont use all of them depending on which dimensions are shown).
> 
> I mean for dimensions that are not default. For example, I have created a new dimension called "ABC" with type slider for project1. ABC does not exist in the list of default dimensions. And now I want to assign dimension ABC to a newly created project2. While we can assign ABC to project2, currently there is no way to add the scale label of ABC to project2. Though it is possible to simply create a new dimension in project2 with the same name, but then we will have two ABC dimensions in db (I thought the idea was that to re-use dimensions that are already created instead of making new dims with same names in every project)

I got you point now, I thought about it previously also but I wasn't sure that we will list all dimensions (default dimensions + newly created dimensions) at the project creation step, that's why I didn't implement it. Also, since generally the project are created by client-api and the label configs can be set by admin. But for this scenario with the UI, we can make it like this: when a new dimension is assigned to a project we will check if there is a corresponded label of that dimension exist in this project, if not a new label with that dimension's name will be added. I will update it to this PR also XD

File: backend/projects/views/dimension.py
Comment (teddy-f-47) [reply backend/projects/views/dimension.py:55]: I was thinking maybe it can be put inside a try-catch block, where if the assigning of dimension1 in `dimension_objs` fails, it will just continue to dimension2 and so on. So if the project already has dims [1, 2, 3], the assigning of [1, 4, 5] will work. What do you think? 😅

**Comment by teddy-f-47** [comment]
> I got you point now, I thought about it previously also but I wasn't sure that we will list all dimensions (default dimensions + newly created dimensions) at the project creation step, that's why I didn't implement it. Also, since generally the project are created by client-api and the label configs can be set by admin. But for this scenario with the UI, we can make it like this: when a new dimension is assigned to a project we will check if there is a corresponded label of that dimension exist in this project, if not a new label with that dimension's name will be added. I will update it to this PR also XD

I think this is a great idea, thanks :)

File: backend/projects/views/dimension.py
Comment (haanh764) [reply backend/projects/views/dimension.py:55]: I think it's better to validate the selected dimension when they select it (for example when they choose 1 there'll be alert that 1 is already exist, pls remove). Because if we let the user pass the whole list of dimensions and let the 4, 5 assigning work silently even the 1 is violated, it's a bit unsafe :sweat_smile: How do u think?

Comment (haanh764) [reply backend/projects/views/dimension.py:55]: ofc this way I will give the endpoint to check the existance :sweat_smile:

Comment (teddy-f-47) [reply backend/projects/views/dimension.py:55]: I see. I think it is reasonable argument. 😅 In this case, let's keep the code as it is now. The validation on FE side can be done because there is already an api for listing the dims of a particular project, which can be used for the purpose. I will mark this point as resolved 😅

**Review by teddy-f-47 — APPROVED** [review]
I have tested the recent changes, everything works fine. I will do a more thorough test again when the FE implementation is ready, but I think this PR can be merged. Thank you for the excellent work! Approved from my side :)

File: backend/projects/views/dimension.py
Comment (teddy-f-47) [inline backend/projects/views/dimension.py]: Sorry, there is a minor change required here 😅. For some weird reason, when I do manual testing, this view works perfectly just fine. However, when using a test script, it throws a 403 error. The reason is apparently missing `permission_classes`. I had to add the following line to make it pass the test script:
```
permission_classes = [IsAuthenticated & (IsProjectAdmin | IsProjectStaffAndReadOnly)]
```

So I hope you can add this line before merging the PR 😅

**Comment by artikandri** [comment]
Hello, thank you for the good work. I have tested it, and I think overall it works fine, but somehow I still managed to add dimension with duplicated name (did it on purpose) even though the API gave me an error message; the new duplicated dimension was still added. I have added the check to prevent duplicate dimension in FE side, of course, but I need to ask: is it intended? 

image

Other than that, everything else works as intended, great work XD

File: backend/projects/views/dimension.py
Comment (haanh764) [reply backend/projects/views/dimension.py]: Yeah thank you XD probably I missed that permission. Added :D

**Comment by haanh764** [comment]
> Hello, thank you for the good work. I have tested it, and I think overall it works fine, but somehow I still managed to add dimension with duplicated name (did it on purpose) even though the API gave me an error message; the new duplicated dimension was still added. I have added the check to prevent duplicate dimension in FE side, of course, but I need to ask: is it intended?
> 
> image
> 
> Other than that, everything else works as intended, great work XD

Hello, it gives you error because of the Label. When adding a new dimension (slider type), a new scale label is added and because of scale label's unique constraints, you got the error message from API. About dimensions with duplicated names, previously I thought that maybe there's the case when we want to have the same dimension name (same question) but with different settings (different metadata codename,..etc). That's why I didn't add the unique constraint on dimension name. But it's oke, to be safer I added the unique constraint on dimension name. It will give you some message like below if you try to add duplicated dimension name
```
{
    "name": [
        "dynamic dimension with this name already exists."
    ],
    "dimension_meta_data": [
        {
            "codename": [
                "dimension meta data with this codename already exists."
            ]
        }
    ]
}
```

**Review by artikandri — APPROVED** [review]
Hello, thank you for the fix XD I have tested it and it works well, approved!

File: backend/projects/views/dimension.py
Comment (teddy-f-47) [inline backend/projects/views/dimension.py:29]: I'm sorry, I just noticed a bug when testing with the FE together with the BE. When I create a new project and assign the dimensions later (basically using the view `AddProjectDimension`, or in other words the `/assign_dimensions` endpoint), the scale label is properly created in that new project.

However, when I create a new project and assign the dimensions in the project creation process (like in the screenshot below -- notice that it is on the project creation page), the scale label for the "testDIM" is not created. Because of this, user can do annotation just fine on the annotation page, but the annotation for that particular dimension won't get saved into DB.
Untitled

Therefore, I think the `perform_create()` method of `DynamicDimensionList` also requires a similar solution of
```
if dimension.id > 69 and dimension.type=='slider':
  ScaleType.objects.create(text=dimension.name, project_id=project_id)
```

Would it be possible?

**Review by teddy-f-47 — CHANGES_REQUESTED** [review]

**Review by teddy-f-47 — APPROVED** [review]
Thank you for the quick fix! I think it should be fine now. Approved once again 😅

File: backend/projects/views/dimension.py
Comment (haanh764) [reply backend/projects/views/dimension.py:29]: Yeah, thank you for finding it. It's my mistake when I added the ability to assign newly created dimension to project creation step, I forgot to update the ability to assign the scale label automatically :sweat_smile: I updated it, result is as below:
Project dimension table:
Screenshot from 2023-03-27 13-39-19

Scale table:
Screenshot from 2023-03-27 13-39-37

**Comment by haanh764** [comment]
Hello @XeonBloomfield @teddy-f-47 @artikandri, I've added the new `package_data_type` flag to project model. Here's the sample of request for project creation with new flag:
```
{   
    "name": "dynamic project test new flag 2",
    "project_type": "DynamicAnnotation",
    "description": "test dynamic final",
    "package_data_type": true,
    "dimension": [{"dimension": [2]}, {"dimension": [13]}, {"dimension": [15]}, {"dimension": [70]}],
    "resourcetype": "DynamicAnnotationProject"
}
```
and result as below:
```
{
    "id": 20,
    "name": "dynamic project test new flag 2",
    "description": "test dynamic final",
    "guideline": "",
    "project_type": "DynamicAnnotation",
    "updated_at": "2023-03-30T15:21:02.280690Z",
    "random_order": false,
    "created_by": 1,
    "collaborative_annotation": false,
    "single_class_classification": false,
    "is_text_project": true,
    "can_define_label": true,
    "can_define_relation": false,
    "can_define_category": true,
    "can_define_span": true,
    "tags": [],
    "allow_overlapping": false,
    "grapheme_mode": false,
    "use_relation": false,
    "is_summary_mode": false,
    "is_emotions_mode": false,
    "is_offensive_mode": false,
    "is_humor_mode": false,
    "is_others_mode": false,
    "is_single_ann_view": true,
    "is_combination_mode": false,
    "package_data_type": true,
    "resourcetype": "DynamicAnnotationProject"
}
```

**Review by teddy-f-47 — APPROVED** [review]
Hello, thank you for the quick update. I have checked the recent change and tested it. The field `package_data_type` has a default value of false, so it works with the current FE implementation without problem. However I have also tested by adding the `package_data_type` field into the request and it works as expected. Hence, accepted from my side. :)

**Pull request merged**
